File: Foundation/iOS/OC/LSCThread.m
//
//  LSCThread.m
//  Sample
//
//  Created by 冯鸿杰 on 2018/12/29.
//  Copyright © 2018年 vimfung. All rights reserved.
//

#import "LSCThread.h"

@interface LSCThread ()

/**
 方法
 */
@property (nonatomic, strong) LSCFunction *_handler;

/**
 执行脚本
 */
@property (nonatomic, strong) LSCScriptController *_scriptController;

@end

@implementation LSCThread

- (instancetype)initWithHandler:(LSCFunction *)handler
{
    if (self = [super init])
    {
        self._handler = handler;
        self._scriptController = [[LSCScriptController alloc] init];
    }
    
    return self;
}

- (void)run:(NSArray<LSCValue *> *)arguments
{
    if (self._handler)
    {
        [self._handler.context runThreadWithFunction:self._handler
                                           arguments:arguments
                                    scriptController:self._scriptController];
    }
}

- (void)exit
{
    [self._scriptController forceExit];
}

@end

File: Foundation/iOS/Swift/LSCThread.swift
//
//  LSCThread.swift
//  Sample
//
//  Created by 冯鸿杰 on 2019/1/2.
//  Copyright © 2019年 vimfung. All rights reserved.
//

#if os(iOS)

import Foundation
import LuaScriptCore_iOS_Swift

#elseif os(OSX)

import Cocoa
import LuaScriptCore_OSX_Swift

#endif

@objc(LSCThread)
class LSCThread: NSObject, LuaExportType
{
    private var _handler : LuaFunction?
    private var _scriptController : LuaScriptController?
    
    @objc init(handler: LuaFunction)
    {
        _scriptController = LuaScriptController();
        _handler = handler;
    }
    
    @objc func run(arguments : Array<LuaValue>) -> Void
    {
        if (_handler != nil)
        {
            var args : Array<LSCValue> = Array<LSCValue>();
            for item in arguments {
                args.append(item.rawValue);
            }
            
            _handler?.context.runThread(with: _handler, arguments: args, scriptController: _scriptController);
        }
    }
    
    @objc func exit() -> Void
    {
        _scriptController?.forceExit();
    }
}
